man/phm2helix.1
.\" -*- NROFF -*-
.\" 
.TH "phm2helix" "1" "" "Kevin Rosenberg and Ian Kay" "Engineering"
.SH "NAME"
phm2helix \- calculate projections through a time varying phantom object.
.SH "SYNOPSIS"
.B phm2helix outfile ndet nview phmprog [OPTIONS]
.SH "DESCRIPTION "
\fIphm2pj\fP calculates projections through a time varying phantom
object.
.SH "OPTIONS"
.TP 16
.B \-\-outfile          
Name of output file for projections
.TP 16
.B \-\-ndet             
Number of detectors
.TP 16
.B \-\-nview            
Number of rotated views
.TP 16
.B \-\-phmfile
Temporary phantom file name
.RS
.TP 16
.B herman        
Herman head phantom
.TP 16
.B herman\-b      
Herman head phantom (Bordered)
.TP 16
.B shepp\-logan   
Shepp\-Logan head phantom
.TP 16
.B shepp\-logan\-b 
Shepp\-Logan head phantom (Bordered)
.TP 16
.B  unitpulse     
Unit pulse phantom
.RE
.TP 16
.B \-\-phmfile \fIname\fP 
Get Phantom from phantom file 
.TP 16
.B \-\-desc           
Description of raysum
.TP 16
.B \-\-nray           
Number of rays per detector (default = 1)
.TP 16
.B \-\-rotangle       
Degrees to rotate view through (multiple of PI) (default = 1)
.TP 16
.B \-\-geometry       
Geometry of scanning
.RS
.TP 16
.B parallel
Parallel scan beams (default)
.TP 16
.B equilinear
Equilinear divergent scan beams
.TP 16
.B equiangular   
Equiangular divergent scan beams
.RE
.TP 16
.B \-\-focal\-length   
Focal length ratio (ratio to radius of phantom) (default = 1)
.TP 16
.B \-\-field\-of\-view  
Field of view (ratio to diameter of phantom square) (default = 1)
.TP 16
.B \-\-trace          
Trace level to use, one of:
.RS 
.TP 
.B none 
No tracing (default)
.TP 16
.B console 
Trace text level
.RE
.TP 16
.B \-\-verbose        
Verbose mode
.TP 16
.B \-\-debug          
Debug mode
.TP 16
.B \-\-version        
Print version
.TP 16
.B \-\-help           
Print a terse help message
.SH "AUTHORS"
Ian Kay, Ph.D.
Kevin Rosenberg, M.D. <kevin@ctsim.org>
.SH "HISTORY"
CTSim was begun in 1983 using MS\-DOS and an EGA display adapter. In
1999 it was ported to GNU/Linux and later ported to Microsoft Windows.
.SH "SEE ALSO"
.BR ctsim (1)
.BR ctsimtext (1)
.BR if1 (1)
.BR if2 (1)
.BR ifexport (1)
.BR ifinfo (1)
.BR linogram (1)
.BR phm2if (1)
.BR phm2pj (1)
.BR pj2if (1)
.BR pjHinterp (1)
.BR pjinfo (1)
.BR pjrec (1)


NEWS
Version 4.2 Released Mar 2003

* Direct Fourier Reconstructions now supported

Version 4.0 Released Jan 2003

* Port to wxWindows 2.4

* Port to Mac OS X


Version 3.5 Released May 2002

* New reconstruction technique: Interpolate divergent beams to
  parallel beams for faster reconstructions.

* Multithreading for running length processes in background. Also
  takes advantage of multiple-CPU's in SMP systems for speed improvements
  with parallel processing.

* Region of interest can be specified for reconstruction zooming.

* Import PNG, PPM, and PGM images

* DICOM file import/export

* Clipboard for cutting and pasting images

* Can read and reconstruct SIEMENS AR.STAR DICOM raw data files

* Scattergram of T-Theta sampling. Useful for understanding divergent beam
  scanning.

* Startup-tips to help new users learn features.

* Ability to place detector at any arbitray position

* Histogram plotting of projection files


Version 3.0 New Features

* Greatly improved dialog boxes

* Online and print manuals with context-sensitive help!

* 3-d rotating views of image files

* Creation of arbitrary filter images

* Cubic interpolation for reconstructions

* All features of command line tools are now in graphical ctsim program!

* Complex-valued image files now supported, lots of image math
 functions added including Fourier transformations

* Visual and statistical image comparision functions

* Plotting of row and column data of single and comparison images

* Histogram Plotting of Image files.

* Conversion of projections to polar images

* Improved animation graphics

* New Microsoft Windows compatible self-installer with explorer extensions 
to CTSim files



Version 2.5 of CTSim has been released!

New Features of CTSim version 2.5:

- Now compiles with Microsoft Visual C++.

- First Microsoft Windows graphical user interface

- Bug fixes and modest documentation and program improvements over version 2.0



New Features of CTSim version 2.0:

- Entire code-base cleaned up, re-written, and converted from C to C++.

- Graphical user interface now included! Animation of projection and
  reconstruction processes. Displays projection and image files. Uses
  platform-independent wxWindows library so can run on UNIX and 
  Microsoft Windows platforms.

- Integrated G2 library for graphical display from command line tools
  (UNIX only)

- Added frequency-based filtering. Can provide significant speed-up
  in reconstructions compared to convolution-based filtering.

- Added highly-optimized backprojection method (idiff3)

- New command-line tools added for image statistics and comparisons with
  both text and graphical outputs.


tools/phm2helix.cpp
/*****************************************************************************
** FILE IDENTIFICATION
**
**   Name:          phm2helix.cpp
**   Purpose:       Take projections of a phantom object
**   Programmers:   Ian Kay and Kevin Rosenberg
**   Date Started:  Aug 2001
**
**  This is part of the CTSim program
**  Copyright (C) 1983-2009 Kevin Rosenberg
**
**  This program is free software; you can redistribute it and/or modify
**  it under the terms of the GNU General Public License (version 2) as
**  published by the Free Software Foundation.
**
**  This program is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**  GNU General Public License for more details.
**
**  You should have received a copy of the GNU General Public License
**  along with this program; if not, write to the Free Software
**  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
******************************************************************************/

#include "ct.h"
#include "timer.h"


enum { O_PHANTOMPROG, O_PHMFILE,  O_DESC, O_NRAY, O_ROTANGLE, O_GEOMETRY, O_FOCAL_LENGTH, O_CENTER_DETECTOR_LENGTH,
  O_VIEW_RATIO, O_SCAN_RATIO, O_OFFSETVIEW,  O_TRACE, O_VERBOSE, O_HELP, O_DEBUG, O_VERSION };

static struct option phm2helix_options[] =
{
  {"phantom", 1, 0, O_PHANTOMPROG},
  {"desc", 1, 0, O_DESC},
  {"nray", 1, 0, O_NRAY},
  {"rotangle", 1, 0, O_ROTANGLE},
  {"geometry", 1, 0, O_GEOMETRY},
  {"focal-length", 1, 0, O_FOCAL_LENGTH},
  {"center-detector-length", 1, 0, O_CENTER_DETECTOR_LENGTH},
  {"view-ratio", 1, 0, O_VIEW_RATIO},
  {"scan-ratio", 1, 0, O_SCAN_RATIO},
  {"offsetview", 1, 0, O_OFFSETVIEW},
  {"trace", 1, 0, O_TRACE},
  {"verbose", 0, 0, O_VERBOSE},
  {"help", 0, 0, O_HELP},
  {"debug", 0, 0, O_DEBUG},
  {"version", 0, 0, O_VERSION},
  {"phmfile", 1, 0, O_PHMFILE},
  {0, 0, 0, 0}
};

static const char* g_szIdStr = "$Id$";


void
phm2helix_usage (const char *program)
{
          std::cout << "usage: " << fileBasename(program) << " outfile ndet nview phmprog [OPTIONS]\n";
          std::cout << "Calculate (projections) through time varying phantom object  \n\n";
          std::cout << "     outfile          Name of output file for projectsions\n";
          std::cout << "     ndet             Number of detectors\n";
          std::cout << "     nview            Number of rotated views\n";
          std::cout << "     phmprog          Name of phm generation executable\n";
          std::cout << "     --phmfile        Temp phantom file name \n";
          std::cout << "     --desc           Description of raysum\n";
          std::cout << "     --nray           Number of rays per detector (default = 1)\n";
          std::cout << "     --rotangle       Angle to rotate view through (fraction of a circle)\n";
          std::cout << "                      (default = select appropriate for geometry)\n";
          std::cout << "     --geometry       Geometry of scanning\n";
          std::cout << "        parallel      Parallel scan beams (default)\n";
          std::cout << "        equilinear    Equilinear divergent scan beams\n";
          std::cout << "        equiangular   Equiangular divergent scan beams\n";
          std::cout << "     --focal-length   Focal length ratio (ratio to radius of phantom)\n";
          std::cout << "                      (default = 1)\n";
          std::cout << "     --view-ratio     Length to view (view diameter to phantom diameter)\n";
          std::cout << "                      (default = 1)\n";
          std::cout << "     --scan-ratio     Length to scan (scan diameter to view diameter)\n";
          std::cout << "                      (default = 1)\n";
          std::cout << "     --offsetview      Intial gantry offset in  'views' (default = 0)\n";
          std::cout << "     --trace          Trace level to use\n";
          std::cout << "        none          No tracing (default)\n";
          std::cout << "        console       Trace text level\n";
          std::cout << "     --verbose        Verbose mode\n";
          std::cout << "     --debug          Debug mode\n";
          std::cout << "     --version        Print version\n";
          std::cout << "     --help           Print this help message\n";
}


int
phm2helix_main (int argc, char* const argv[])
{
        Phantom phm;
        std::string optGeometryName = Scanner::convertGeometryIDToName(Scanner::GEOMETRY_PARALLEL);
        char *opt_outfile = NULL;
        std::string opt_desc;
        std::string opt_PhmProg;
        std::string opt_PhmFileName = "tmpphmfile";
        int opt_ndet;
        int opt_nview;
        int opt_offsetview = 0;
        int opt_nray = 1;
        double dOptFocalLength = 2.;
        double dOptCenterDetectorLength = 2;
        double dOptViewRatio = 1.;
        double dOptScanRatio = 1.;
        int opt_trace = Trace::TRACE_NONE;
        int opt_verbose = 0;
        int opt_debug = 0;
        double opt_rotangle = -1;
        char* endptr = NULL;
        char* endstr;

        Timer timerProgram;

        while (1) {
                int c = getopt_long(argc, argv, "", phm2helix_options, NULL);

                if (c == -1)
                break;

        switch (c) {
        case O_VERBOSE:
                opt_verbose = 1;
                break;
        case O_DEBUG:
                opt_debug = 1;
                break;
        case O_TRACE:
                if ((opt_trace = Trace::convertTraceNameToID(optarg))
                                                                == Trace::TRACE_INVALID) {
                        phm2helix_usage(argv[0]);
                        return (1);
                }
                break;
                          case O_PHMFILE:
                                opt_PhmFileName = optarg;
                                break;
                          case O_DESC:
                                opt_desc = optarg;
                                break;
                          case O_ROTANGLE:
                                opt_rotangle = strtod(optarg, &endptr);
                                endstr = optarg + strlen(optarg);
                                if (endptr != endstr) {
                                        std::cerr << "Error setting --rotangle to " << optarg << std::endl;
                                        phm2helix_usage(argv[0]);
                                        return (1);
                                }
                                break;
                          case O_GEOMETRY:
                                optGeometryName = optarg;
                                break;
                          case O_FOCAL_LENGTH:
                                dOptFocalLength = strtod(optarg, &endptr);
                                endstr = optarg + strlen(optarg);
                                if (endptr != endstr) {
                                        std::cerr << "Error setting --focal-length to " << optarg << std::endl;
                                        phm2helix_usage(argv[0]);
                                        return (1);
                                }
                                break;
                          case  O_CENTER_DETECTOR_LENGTH:
                                dOptCenterDetectorLength = strtod(optarg, &endptr);
                                endstr = optarg + strlen(optarg);
                                if (endptr != endstr) {
                                        std::cerr << "Error setting --center-detector-length to " << optarg << std::endl;
                                        phm2helix_usage(argv[0]);
                                        return (1);
                                }
                          break;
                          case O_VIEW_RATIO:
                                dOptViewRatio = strtod(optarg, &endptr);
                                endstr = optarg + strlen(optarg);
                                if (endptr != endstr) {
                                        std::cerr << "Error setting --view-ratio to " << optarg << std::endl;
                                        phm2helix_usage(argv[0]);
                                        return (1);
                                }
                                break;
                          case O_SCAN_RATIO:
                                dOptScanRatio = strtod(optarg, &endptr);
                                endstr = optarg + strlen(optarg);
                                if (endptr != endstr) {
                                        std::cerr << "Error setting --scan-ratio to " << optarg << std::endl;
                                        phm2helix_usage(argv[0]);
                                        return (1);
                                }
                                break;
                          case O_NRAY:
                                opt_nray = strtol(optarg, &endptr, 10);
                                endstr = optarg + strlen(optarg);
                                if (endptr != endstr) {
                                  std::cerr << "Error setting --nray to %s" << optarg << std::endl;
                                  phm2helix_usage(argv[0]);
                                  return (1);
                                }
                                break;
                          case O_OFFSETVIEW:
                                opt_offsetview = strtol(optarg, &endptr, 10);
                                endstr = optarg + strlen(optarg);
                                if (endptr != endstr) {
                                  std::cerr << "Error setting --offsetview to %s" << optarg << std::endl;
                                  phm2helix_usage(argv[0]);
                                  return (1);
                                }
                                break;
                          case O_VERSION:
#ifdef VERSION
                                  std::cout << "Version: " << VERSION << std::endl << g_szIdStr << std::endl;
#else
                                          std::cout << "Unknown version number\n";
#endif
                                  return (0);
                          case O_HELP:
                          case '?':
                                phm2helix_usage(argv[0]);
                                return (0);
                          default:
                                phm2helix_usage(argv[0]);
                                return (1);
                        } // end of switch
          } // end of while loop

          if (optind + 4 != argc) {
                phm2helix_usage(argv[0]);
                return (1);
          }

          opt_outfile = argv[optind];
          opt_ndet = strtol(argv[optind+1], &endptr, 10);
          endstr = argv[optind+1] + strlen(argv[optind+1]);
          if (endptr != endstr) {
                std::cerr << "Error setting --ndet to " << argv[optind+1] << std::endl;
                phm2helix_usage(argv[0]);
                return (1);
          }
          opt_nview = strtol(argv[optind+2], &endptr, 10);
          endstr = argv[optind+2] + strlen(argv[optind+2]);
          if (endptr != endstr) {
                std::cerr << "Error setting --nview to " << argv[optind+2] << std::endl;
                phm2helix_usage(argv[0]);
                return (1);
          }
          opt_PhmProg = argv[optind+3];

          if (opt_rotangle < 0) {
                if (optGeometryName.compare ("parallel") == 0)
                  opt_rotangle = 0.5;
                else
                  opt_rotangle = 1.0;
          }

          std::ostringstream desc;
          desc << "phm2helix: NDet=" << opt_ndet
                   << ", Nview=" << opt_nview
                   << ", NRay=" << opt_nray
                   << ", RotAngle=" << opt_rotangle
                   << ", OffsetView =" << opt_offsetview
                   << ", Geometry=" << optGeometryName
                   << ", PhantomProg=" << opt_PhmProg
                   << ", PhmFileName=" << opt_PhmFileName;
          if (opt_desc.length()) {
                desc << ": " << opt_desc;
          }
          opt_desc = desc.str();

          opt_rotangle *= TWOPI;

          int stat;
          char extcommand[100];
          if(opt_debug != 0)
                        std::cout  <<  opt_PhmProg  <<  " " << 0 << " " <<  opt_nview << " " << opt_PhmFileName  << std::endl;
           //extcommand <<  opt_PhmProg  <<  " " << 0 << " " <<  opt_nview << " " << opt_PhmFileName ;

          sprintf(extcommand, "%s %d %d %s",    opt_PhmProg.c_str(), 0, opt_nview, opt_PhmFileName.c_str() );

           stat = system( extcommand );
          if (stat != 0 )
                        std::cerr << "Error executing external phantom program " << opt_PhmProg << " with command " << extcommand << std::endl;

          phm.createFromFile (opt_PhmFileName.c_str());
          remove(opt_PhmFileName.c_str());

          Scanner scanner (phm, optGeometryName.c_str(), opt_ndet, opt_nview,
                                opt_offsetview, opt_nray, opt_rotangle, dOptFocalLength,
                                dOptCenterDetectorLength, dOptViewRatio, dOptScanRatio);
          if (scanner.fail()) {
                 std::cout << "Scanner Creation Error: " << scanner.failMessage()
                                                        << std::endl;
                 return (1);
          }

          Projections pjGlobal(scanner);


          for( int iView = 0; iView < opt_nview; iView++ ){
                if(opt_debug != 0)
                        std::cout  <<  opt_PhmProg  <<  " " << iView << " " <<  opt_nview << " " << opt_PhmFileName  << std::endl;
           //extcommand <<  opt_PhmProg  <<  " " << iView << " " <<  opt_nview << " " << opt_PhmFileName ;

                sprintf(extcommand, "%s %d %d %s",      
                        opt_PhmProg.c_str(), iView, opt_nview, 
                        opt_PhmFileName.c_str() );
                stat = system( extcommand );

                if (stat != 0 )
                        std::cerr << "Error executing external phantom program " << opt_PhmProg << " with command " << extcommand << std::endl;
                Phantom phmtmp;
                phmtmp.createFromFile (opt_PhmFileName.c_str());

                scanner.collectProjections (pjGlobal, phmtmp, iView,
                          1, scanner.offsetView(), true, opt_trace);
                remove(opt_PhmFileName.c_str());
          }


          pjGlobal.setCalcTime (timerProgram.timerEnd());
          pjGlobal.setRemark (opt_desc);
          pjGlobal.write (opt_outfile);
          if (opt_verbose) {
            phm.print (std::cout);
            std::cout << std::endl;
            std::ostringstream os;
            pjGlobal.printScanInfo (os);
            std::cout << os.str() << std::endl;
            std::cout << "  Remark: " << pjGlobal.remark() << std::endl;
            std::cout << "Run time: " << pjGlobal.calcTime() << " seconds\n";
          }
          
          return (0);
}


#ifndef NO_MAIN
int
main (int argc, char* argv[])
{
  int retval = 1;

  try {
    retval = phm2helix_main(argc, argv);
#if HAVE_DMALLOC
    //    dmalloc_shutdown();
#endif
  } catch (exception e) {
    std::cerr << "Exception: " << e.what() << std::endl;
  } catch (...) {
    std::cerr << "Unknown exception\n";
  }

  return (retval);
}
#endif

